### PR TITLE
Add support to skip action notifications when executed under workflow context

### PR DESCRIPTION
**Pull request by khushboobhatia01**
This pull request resolves https://github.com/StackStorm/st2/issues/5221

### CODE REVIEW

File: contrib/runners/orquesta_runner/tests/unit/test_notify.py
Comment (Kami) [inline contrib/runners/orquesta_runner/tests/unit/test_notify.py:353]: If we don't have test cases for various other scenarios already (multiple tasks in a workflow, notify set only for some of them, all of them, etc.), it would be good to add them.

If we already have them, please ignore my comment.

File: st2common/st2common/services/action.py
Comment (Kami) [inline st2common/st2common/services/action.py:561]: I believe this will only work / correctly detect Orquesta workflows, but not action-chain ones, right?

Does this issue / bug only affect Orquesta workflows or also action-chain ones? If it also affects action chain ones, we probably also need to handle that scenario here

**Comment by Kami** [comment]
Thanks for the contribution.

The bug fix / change looks good to me overall, I just had some questions / comments as far as tests and action-chain workflows go.

File: st2common/st2common/services/action.py
Comment (Kami) [inline st2common/st2common/services/action.py:563]: Just thinking out loud.

This means that we will now perform 2 additional DB queries for any workflow action when calling ``create_request``, right?

I guess that is probably not that big deal, but it may make sense to utilize "only_fields" so we only retrieve fields we need on those two models (``notify`` for ``WorkflowExecutionDB``, ``task_name`` for ``TaskExecution``).

---

EDIT: Yeah, ``WorkflowExecutionDB`` can also contain ``result`` and other large fields values so it is actually important we only retrieve fields we need.

Luckily my recent DB performance optimizations make this less of an issue (in the past, retrieving large fields such as results from the database could easily take many seconds or more), but we should still be careful and avoid retrieving unnecessary fields / data we don't need.

File: st2common/st2common/services/action.py
Comment (khushboobhatia01) [reply st2common/st2common/services/action.py]: Verified, issue exists for action-chain workflows as well. Modified the changes to fix same for action-chains. The fix isn't as direct as it for orquesta, as skip_notify parameter is available only in runner parameters. Please review and let me know if the changes look ok to you. Keeping this thread unresolved till then.

Comment (Kami) [reply st2common/st2common/services/action.py]: Thanks for also tackling this for action chain :+1:

Comment (khushboobhatia01) [reply st2common/st2common/services/action.py]: @Kami If the changes look ok to you, can you please approve and merge? Thanks :)

File: contrib/runners/orquesta_runner/tests/unit/test_notify.py
Comment (khushboobhatia01) [reply contrib/runners/orquesta_runner/tests/unit/test_notify.py]: Addressed.

File: st2common/st2common/services/action.py
Comment (khushboobhatia01) [reply st2common/st2common/services/action.py]: Addressed.

**Review by Kami — APPROVED** [review]
LGTM :+1: 

Thanks for working.

I will merge this into master and add changelog entry separately so I don't block and hold this change any longer.

**Comment by Kami** [comment]
For some reason after merging this, those new tests are failing for me locally with this error - https://gist.github.com/Kami/52543866f5400e29537f00911654bfe1.

No idea (yet) why it happens locally, but not on CI. I likely won't have time to dig in before the weekend.

**Comment by Kami** [comment]
OK, I see the issue - https://github.com/StackStorm/st2/pull/5227/files#diff-c86e8df3fb95a5e1b0806ce20a77ee1a92f721e933a1d230e5a302b1ebb10873R210.

This function needs some more love. Right now it will return as soon as execution has 1 or more children, but from the code that's not actually what we want - we want to wait for 2 children. So the code needs to be changed to handle that better otherwise it will be prone to timing / race conditions.

I will push a fix shortly.

**Comment by Kami** [comment]
Here we go - 5856fa482.

I confirmed that with this change, the tests now consistently pass locally.

**Comment by khushboobhatia01** [comment]
> Here we go - [5856fa4](https://github.com/StackStorm/st2/commit/5856fa482f3371696ca0574b4576722c34ee3e8d).
> 
> I confirmed that with this change, the tests now consistently pass locally.

Thanks for fixing the test case 👍

**Pull request merged**
